_posts/2016-02-25-on-idempotency-in-distributed-rest-apis.md

```diff
@@ -0,0 +1,132 @@
+---
+layout: post
+title: "Idempotency in REST APIs and distributed systems"
+ghavatar: 310118
+ghname: gronnbeck
+jobtitle: Platform Engineer
+tags: [rest, api, distributed systems]
+---
+
+A few weeks back our team had a discussion about API design. The discussion was
+mostly about what we could do to design pragmatic and user friendly RESTful APIs for our clients.
+At some point in the discussion we discussed what idempotency is and what it
+actually guarantees. That sparked a curiosity in me, and I wanted to explore deeper
+into that.
+
+Idempotency is defined as: ([wikipedia](https://en.wikipedia.org/wiki/Idempotence)):
+
+> Idempotence is the property of certain operations in mathematics and computer
+> science, that can be applied multiple times without changing the result beyond
+> the initial application.
+
+That is, a HTTP method is idempotent if it guarantees that repeating a
+request multiple of times has the same effect as issuing the request once.
+This is especially important in the case of network failures. In such cases
+clients can repeat the same request multiple times without worrying about
+any unintended effects.
+
+## Safe vs Idempotent
+[Idempotent methods](https://tools.ietf.org/html/rfc7231#section-4.2.2)
+should not be confused with
+[Safe methods](https://tools.ietf.org/html/rfc7231#section-4.2.1).
+In HTTP, method is safe if they are not expected to cause side effects.
+Meaning clients can send request to safe methods without worrying about
+causing any side effects or changes to the resource.
+
+The table below gives an overview of the different properties for the most
+popular HTTP methods used when implementing RESTful APIs,
+``GET``, ``PUT``, ``POST`` and ``DELETE``.
+
+| Method | Safe | Idempotent |
+|:-------|:----:|:----------:|
+| GET    | Yes  | Yes        |
+| PUT    | No   | Yes        |
+| POST   | No   | No         |
+| DELETE | No   | Yes        |
+
+
+## Pragmatic vs Idempotency
+
+Being idempotent is important. However, it does not mean that you
+have to guarantee that a request one point
+in time should always return the same result. E.g. applying ``DELETE`` on
+a existing resource multiple times should return ``HTTP Status: OK (200)``
+the first time and ``HTTP Status: Not Found (404)`` the subsequent times.
+
+A strict implementation of ``DELETE`` would expect ``HTTP Status: OK (200)``.
+I believe that approach to be unnecessary cumbersome and pragmatism should
+trump idempotency in such cases.
```

Comment:
Well, in its isolated form PUT can return the same status code if the input is exactly the same. That is if we ignore the "optimistic locking" part later on.